components/TextFields/src/MDCTextInputBehavior.h

```diff
@@ -0,0 +1,174 @@
+   count / limit below text.
+   */
+  MDCTextInputPresentationStyleDefault = 0,
+
+  /**
+   The placeholder text is laid out inline but will float above the field when there is content or
+   the field is being edited. The character count is below text. The Material Design guidelines
+   call this 'Floating inline labels.'
+   https://material.io/guidelines/components/text-fields.html#text-fields-labels
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/**
+ Material Design themed states for textInputs. The logic for 'automagic' error states changes:
+ underline color, underline text color.
+ https://www.google.com/design/spec/components/text-fields.html#text-fields-single-line-text-field
+ */
+@interface MDCTextInputBehavior : NSObject <NSCoding, NSCopying>
+
+/** The text input the behavior is managing. */
+@property(nonatomic, nullable, weak) UIView<MDCTextInput> *textInput NS_SWIFT_NAME(input);
+
+/**
+ The color used to denote error state in the underline, the errorText's label, the plceholder and
+ the character count label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, strong) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ Controls when the underline will be shown.
+
+ The underline is an overlay that can be hidden depending on the editing state of the input text.
+
+ Default is UITextFieldViewModeAlways.
+ */
+@property(nonatomic, assign) UITextFieldViewMode underlineViewMode NS_SWIFT_NAME(underlineMode)
+    UI_APPEARANCE_SELECTOR;
+
+/**
+ The character counter. Override to use a custom character counter.
+
+ Default is an internal instance MDCTextInputAllCharactersCounter. Setting this property to null 
+ will reset it to return that instance.
+ */
+@property(nonatomic, null_resettable, weak) IBInspectable id<MDCTextInputCharacterCounter>
+    characterCounter;
+
+/**
+ Controls when the character count will be shown and therefore whether character counting determines
+ error state.
+
+ Default is UITextFieldViewModeNever.
+ */
+@property(nonatomic, assign) UITextFieldViewMode characterCountViewMode NS_SWIFT_NAME(characterMode)
+    ;
+
+/**
+ The character limit for the text input. A label under the input counts characters entered and
+ presents the count / the limit.
+
+ If character count / limit has been hidden by the characterCountViewMode
+ (ie: UITextFieldViewModeNever) changing the value of characterLimit has no effect.
+
+ If the character count goes above its limit, the underline, the character count / limit label and
+ any floating placeholder label all turn to the error color; the text input will be in error state.
+ Note: setErrorText:errorAccessibilityValue: also sets these MDCTextInput properties.
+ 
+ Default is 0.
+ */
+@property(nonatomic, assign) IBInspectable NSUInteger characterLimit;
```

Comment:
we should match names with the other property and make it clear its an upper limit. characterCountMax
which also brings up a idea... how does one do a minimum?